### DISCUSSION TITLE
OQRS search in all station locations

### DISCUSSION THREAD

**Discussion by R1BLH** [magicbug/Cloudlog#1906] [Ideas / Feature Requests]
It would be good to make OQRS search for QSO in all station locations of the one logbook (callsign).
For example. I have 25 station locations in the my logbook now: https://r1blh.cats-home.net/log/index.php/oqrs
It would be more comfortable for user to make one search in whole logbook (like on the ClubLog), not many searches in different station locations.

**Comment by AndreasK79** [comment]
OQRS was written for each station location being it's own log in mind. That's usually the reason you have different station locations.
Any particular reason you have several station locations on the same call and same qth?

**Reply by R1BLH** [reply]
>Any particular reason you have several station locations on the same call and same qth?

@AndreasK79 
I have two basic QTH, one of them is typically active on weekends in warm season (my countryside house). In order to use eQSL i need make new station location for each time period for each QTH (it is eQSL.cc rule, so in each station location data contains QTH name from eQSL.cc, the same as station location name). 
You could see this from naming of them:
"230131-221115-KP50da" means "November 22 2022 - January 31 2023, at home (KP50da)"
"221113-221112-KO59ef" means "November 12 2022 - November 13 2022, at countryside (KO59ef)"
"220807-220806-KP40qj" means "August 6 2022 - August 7 22, somewhere in a forest (KP40qj)"

I would be happy to have small set of QTH which would be depend from location, not from time periods, but eQSL rules prevents this.
So this is a reason to search QSO in whole log.

**Reply by AndreasK79** [reply]
@cats-shadow thanks for the explanation. Now I understand the reason behind it.

I make no promises, but I'll give it some thought and find some solution.

**Reply by R1BLH** [reply]
@AndreasK79 thank you!
May be it would be simple solution to add an option search QSO from whole table "TABLE_HRD_CONTACTS_V01"?

**Reply by AndreasK79** [reply]
That will not work. OQRS was written to use each location as it's own log. I tried to remove station_id, and the first problem was that band/mode table is not correct, since that is also based on station_id.

If I continue, the table where you fill inn QSOs does now show which callsign you request from, so in case of duplicates between locations, you would not know which QSO you requested from.

I stopped there, as it was already broken enough at this point.

This is not something that I will do anything about now. I might pick it up again at a later point, but no promises.

**Reply by R1BLH** [reply]
> This is not something that I will do anything about now. I might pick it up again at a later point, but no promises.

Thank you.
Now in CloudLog we have two entity arrays: "station logbook" and "station location", and each "station logbook" can contain multiple "station location" . The element names of both arrays are unknown to anyone who does not own a given ClowdLog instance, so the requestor can only make a guess at which "station location" to request a QSO, except that they can more likely guess at which "station logbook" to look for a QSO.

The main idea of my question was to make OQRS convenient for the requesting party. The requestor usually does not know what the "station location" should be called, but he knows his own QSO data: 
its own call sign -- COL_CALL
call sign -- COL_STATION_CALLSIGN
date/time -- COL_TIME_ON, COL_TIME_OFF 
range -- COL_BAND, 
mode -- COL_MODE
The set of these data is unique for each QSO and if it is duplicated in several "station locations", it is still the same QSO. It is physically impossible for an operator to make several QSOs with the same type of radiation with the same other station in several locations at the same time.
I now tried to do a SQL query in the table TABLE_HRD_CONTACTS_V01 with a search for one Callsign, which I had a few QSOs (as if you enter only your Callsign in the query OQRS -- SELECT * FROM `TABLE_HRD_CONTACTS_V01` WHERE COL_CALL = "4U1A" 
), got the correct answer -- a few lines.
This is roughly how OQRS works in Clublog. You enter the Callsign, get the result table with the number of contacts for each range of each mode, and then refine the QSL query by entering the date and time.
Thank you for patience!

**Reply by AndreasK79** [reply]
I do understand what you mean.

To be able to do this, a lot of stuff needs to be rewritten. You can't just search the whole log table (by removing station_id in query), then present the results.

Perhaps OQRS should have been written in another way, but I did it in a way I thought would be logical. 
Since people are using Cloudlog in different ways, OQRS is not working optimal for everyone.

**Reply by R1BLH** [reply]
Copy that.
Sorry to bother you. I was just trying to look at the OQRS process from the perspective of the person who would be requesting the QSL.

I tried looking at the code, it's a bit difficult for me since I'm not a web programmer (if I do write code, it's either PC applications or device firmware).

**Reply by AndreasK79** [reply]
I appreciate the feedback, so no need to be sorry :)

Yes, it would be easier for a user to search all logs, but it need to be done in a way that works. 

Let's say you search all logs. How would you present it? A table with bands and modes would work, but only if the station locations all had the same callsign. If not, you wouldn't know for which call the table was for.

If all station locations were to be searched at once, I guess skipping this table, and instead listing QSOs would be better. 
The user could then fill in information and request from there.

**Reply by R1BLH** [reply]
> Let's say you search all logs. How would you present it? A table with bands and modes would work, but only if the station locations all had the same callsign. If not, you wouldn't know for which call the table was for.

Roger! :)

Of course, the call sign for which the OQRS request is made must be the same.
I will try to write a possible search algorithm, as I see it.
If in CloudLog contains locations of stations with different callsigns, you can make a table from which the user will choose (dropdownlist, for example) for which of the callsigns in CloudLog it will make an OQRS request.
This table in the database can be formed as the new stations locations are entered, so you do not have to run a query every time you request OQRS. 
After the user selects a Callsign, for which will be forming an OQRS from this table, can be obtained a list of station locations for the selected Callsign, and for each of the station in this list perform an existing procedure for searching QSOs.
The search result can be merged into one mode/band table, while the search result itself can be stored in an array with one additional field "station location" (I don't know how it is stored now).

> 
> If all station locations were to be searched at once, I guess skipping this table, and instead listing QSOs would be better. The user could then fill in information and request from there.

If the user wants to search QSOs for all callsigns for stations in CloudLog, in fact it turns out one more cycle, within which the algorithm described above will be executed. Result would be presented as different tables for each Callsign.

**Comment by AndreasK79** [comment]
@cats-shadow I've added grouped search now. You can enable it under admin -> global options -> oqrs.
This is currently in the dev branch.

**Reply by R1BLH** [reply]
Thank you!

**Reply by R1BLH** [reply]
Hmm. I found a strange behavior on my installation. If "Grouped search" is of  there is no field to enter the callsign on the OQRS page. Please look on the picture of page after station location has been selected.
Could you check on yours install, please? (may be this is local bug on my install?)

![2023-05-11_00-01-42](https://github.com/magicbug/Cloudlog/assets/16667173/0917432b-8873-4908-82c6-5bb1bef2c3b8)

**Reply by AndreasK79** [reply]
It might be that you need to clear your cache. The search does work on your page now.
